rootfs/etc/nginx/lua/test/configuration_test.lua

```diff
@@ -0,0 +1,114 @@
+            assert.spy(s).was_called_with("Only POST requests are allowed!")
+            assert.same(ngx.status, ngx.HTTP_BAD_REQUEST)
+        end)
+
+        it("should ignore servers that don't have hostname or pemCertKey set", function()
+            ngx.var.request_method = "POST"
+            local mock_servers = cjson.encode({
+                {
+                    hostname = "hostname",
+                    sslCert = {}
+                },
+                {
+                    sslCert = {
+                        pemCertKey = "pemCertKey"
+                    }
+                }
+            })
+            ngx.req.get_body_data = function() return mock_servers end
+
+            local s = spy.on(ngx, "log")
+            assert.has_no.errors(configuration.handle_servers)
+            assert.spy(s).was_called_with(ngx.WARN, "hostname or pemCertKey are not present")
+            assert.same(ngx.status, ngx.HTTP_CREATED)
+        end)
+
+        it("should successfully update certificates and keys for each host", function()
+            ngx.var.request_method = "POST"
+            local mock_servers = cjson.encode({
+                {
+                    hostname = "hostname",
+                    sslCert = {
+                        pemCertKey = "pemCertKey"
+                    }
+                }
+            })
+            ngx.req.get_body_data = function() return mock_servers end
+
+            assert.has_no.errors(configuration.handle_servers)
+            assert.same(ngx.status, ngx.HTTP_CREATED)
+        end)
+
+        it("should log an err and set status to Internal Server Error when a certificate cannot be set", function()
+            ngx.var.request_method = "POST"
+            ngx.shared.certificate_data.safe_set = function(self, data) return false, "error" end
+            local mock_servers = cjson.encode({
+                {
+                    hostname = "hostname",
+                    sslCert = {
+                        pemCertKey = "pemCertKey"
+                    }
+                }
+            })
+            ngx.req.get_body_data = function() return mock_servers end
+
+            local s = spy.on(ngx, "log")
+            assert.has_no.errors(configuration.handle_servers)
+            assert.spy(s).was_called_with(ngx.ERR, "error setting certificate for hostname: error\n")
+            assert.same(ngx.status, ngx.HTTP_INTERNAL_SERVER_ERROR)
+        end)
+
+        it("should log an err and set status to Internal Server Error when shared dictionary is full", function()
+            ngx.var.request_method = "POST"
+            ngx.shared.certificate_data.safe_set = function(self, data) return false, "no memory" end
+            local mock_servers = cjson.encode({
+                {
+                    hostname = "hostname",
+                    sslCert = {
+                        pemCertKey = "pemCertKey"
+                    }
+                }
+            })
```

Comment:
can you have more than one server and return "no memory" on only i.e second of them and assert that the loop gets broken and function does not try to store the next one